### PR TITLE
Add label merging function (and improve ndmeasure.label)

### PR DESCRIPTION
**Pull request by m-albert**
## PR

This PR contains work I've been wanting to discuss and push here for a while.

Highlights:
1. Improvements to `ndmeasure.label`:
    1. Implementation of the idea by @jakirkham and @jni to make labels unique by using the lower and upper bits of the label array (see [here](https://github.com/dask/dask-image/issues/199)). This improves parallelism at the expensive of returning non-contiguous labels.
    1. The current implementation of `ndmeasure.label` doesn't output contiguous (sequential) labels (also before implementing the point above). This PR adds a parameter `produce_sequential_labels` to `dask_image.ndmeasure.label` to output contiguous labels.
1. Implementation of a chunked version of `skimage.segmentation.relabel_sequential`: `ndmeasure.relabel_sequential` (this provides the functionality e.g. of the point above, happy to discuss whether the function should be public)
1. Implementation of `ndmeasure.merge_labels_across_chunk_boundaries` (details below)
    1. Use case 1: merge touching labels
    2. Use case 2: merge labels above a given IoU within the overlap.

## merge_labels_across_chunk_boundaries
What's the idea behind `ndmeasure.merge_labels_across_chunk_boundaries`?

### Visual abstract

Cases of overlap = 0 (top) and overlap > 0 (bottom)

image

### Description

From the docstring:

```python
def merge_labels_across_chunk_boundaries(
        labels,
        overlap_depth=0,
        iou_threshold=0.8,
        structure=None,
        wrap_axes=None,
        trim_overlap=True,
        produce_sequential_labels=True,
        ):
    """
    Merge labels across chunk boundaries.

    Each chunk in ``labels`` has been labeled independently, and the labels
    in different chunks may overlap. This function tries to merge labels across
    chunk boundaries using a strategy dependent on ``overlap``:
    - If ``overlap > 0``, the overlap region between chunks is used to
      determine which between each pair of chunks should be merged.
    - If ``overlap == 0``, labels that touch across chunk boundaries are merged.

    Parameters
    ----------
    labels : dask array of int
        The input labeled array, where each chunk is independently labeled.
    overlap_depth : int, optional
        The size of the overlap region between chunks, e.g. as produced by
        `dask.array.overlap` or `map_overlap`. Default is 0.
    iou_threshold : float, optional
        If ``overlap > 0``, the intersection-over-union (IoU) between labels
        in the overlap region is used to determine which labels should be
        merged. If the IoU between two labels is greater than ``iou_threshold``,
        they are merged. Default is 0.8.
    structure : array of bool, optional
        Structuring element for determining connectivity. If None, a
        cross-shaped structuring element is used.
    wrap_axes : tuple of int, optional
        Should labels be wrapped across array boundaries, and if so which axes.
        - (0,) only wrap over the 0th axis.
        - (0, 1) wrap over the 0th and 1st axis.
        - (0, 1, 3)  wrap over 0th, 1st and 3rd axis.
    trim_overlap : bool, optional
        If True, the overlap regions are trimmed from the output labels.
        Default is True.
    """
```

### Example

Consider the following problem.

The user wants to segment a large input image, i.e. a dask array.

```python
import numpy as np
import dask.array as da
import skimage
import matplotlib.pyplot as plt

dim = da.from_array(
    skimage.data.cells3d()[:, 1, :, :].max(axis=0), chunks=(110, 110))
```
image

and has a segmentation method that cannot be applied on the entire image (e.g. memory or method constraints) but works well on regions of the input image. The user can use `map_overlap` to apply the segmentation on the chunks of the input image and include overlap so that the segmentation method performs well on the chunk boundaries:

```python
model = models.Cellpose(gpu=False, model_type='cyto')

dseg = dim.map_overlap(
        lambda x: model.eval(x, diameter=None, channels=[0, 0])[0],
        dtype=np.uint16,
        depth=10,
    ).persist()

plt.figure()
plt.imshow(dseg)
```
image

At this point, the user faces two challenges:
1. the obtained segmentation labels show boundary artefacts and
3. they are not unique.

This is where `ndmeasure.merge_labels_across_chunk_boundaries` comes in to merge labels across chunk boundaries.

#### First use case / usage pattern
Similarly to the implementation of `ndmeasure.label` and e.g. the approach provided in [distributed cellpose](https://github.com/MouseLand/cellpose/pull/1062) (ping @GFleishman), one approach for merging labels consists in merging labels that touch across chunk boundaries. `ndmeasure.merge_labels_across_chunk_boundaries` allows to do this in a way that's agnostic of the segmentation method:

```python
import dask_image.ndmeasure

dseg_merged = dask_image.ndmeasure.merge_labels_across_chunk_boundaries(
    dseg
)['labels']

plt.figure()
plt.imshow(dseg_merged)
```

image

#### Second use case

The approach above joins all labels that form part of the same connected component in a boundary slice of thickness one. This can lead to artefacts. In fact, the example image above contains two merged cells which are clearly separated in the chunkwise segmentation. To separate these, `dask_image.ndmeasure.merge_labels_across_chunk_boundaries` has the option to merge labels based on the IoU within an overlap region.

```python
dseg = da.overlap.overlap(
        dim, depth={0: 10, 1: 10},
        boundary='none'
    ).map_blocks(
        lambda x: model.eval(x, diameter=None, channels=[0, 0])[0],
        dtype=np.uint16
    ).persist()

dseg_merged = dask_image.ndmeasure.merge_labels_across_chunk_boundaries(
    dseg, overlap_depth=10, iou_threshold=0.8,
)['labels'].compute(scheduler='single-threaded')

plt.figure()
plt.imshow(dseg_merged)
```
image

The labels of the two cells below which previously had been joined now remain separated.

## Finally

In my personal projects it's been useful to have a function that merges labels across chunks. I think `dask-image` could be a good place for providing such functionality to the community and I've had some offline feedback at meetings that this could be useful. Happy about any further feedback / discussion over here!

### CODE REVIEW

**Comment by m-albert** [comment]
Thanks @jni 🤗 Glad you like it and that you had the same idea!

Regarding this PR I still need to get an (obscurely) failing test under control...
